File: day2/swift/main.swift
// AoC day 2

import Foundation

extension String {
    func numberOfOccurances(string: String) -> Int {
        return self.components(separatedBy: string).count - 1
    }
    subscript(offset: Int) -> Character {
        self[index(self.startIndex, offsetBy: offset)]
    }
}

class passwordRule {
    var minOccurance: Int = 0
    var maxOccurance: Int = 0
    var pwdChar: String = ""
    
    func Validate(s: String) -> Bool {
        let occurances = s.numberOfOccurances(string: pwdChar)
        let minOk = occurances >= minOccurance
        let maxOk = occurances <= maxOccurance
        return minOk && maxOk
    }
    func Validate2(s: String) -> Bool {
        let c1 = s[minOccurance - 1]
        let c2 = s[maxOccurance - 1]
        let ok1 = String(c1) == pwdChar
        let ok2 = String(c2) == pwdChar
        let bothOk = ok1 && ok2
        return (ok1 || ok2) && !bothOk
    }
}

func parseRule(data: String) -> (rule: passwordRule, password: String) {
    let result = passwordRule()
    let parts = data.split(separator: " ")
    let ranges = parts[0].split(separator: "-")
    if let r1 = Int(ranges[0]) {
        result.minOccurance = r1
    }
    if let r2 = Int(ranges[1]) {
        result.maxOccurance = r2
    }
    result.pwdChar = parts[1].replacingOccurrences(of: ":", with: "")
    
    return (result, String(parts[2]))
}

func readFile(filename: String) -> [String] {
    do {
        let text = try String(contentsOfFile: filename, encoding: .utf8)
        let lines : [String] = text.components(separatedBy: "\n")
        return lines
    } catch let error as NSError{
        return[]
    }
}

func solve1() {
    let rawData = readFile(filename: "/Users/mikaelstalvik/Desktop/aoc/AoC2/aoc2.txt")
    
    var rules: [passwordRule] = []
    var passwords: [String] = []
    for item in rawData {
        if (!item.isEmpty) {
            let (rule, pwd) = parseRule(data: item)
            rules.append(rule)
            passwords.append(pwd)
        }
    }
    var res = 0
    var i = 0
    for password in passwords {
        if (rules[i].Validate(s: password)) {
            res += 1
        }
        i += 1
    }
    print("Valid: \(res)")
}

func solve2() {
    let rawData = readFile(filename: "/Users/mikaelstalvik/Desktop/aoc/AoC2/aoc2.txt")
    
    var rules: [passwordRule] = []
    var passwords: [String] = []
    for item in rawData {
        if (!item.isEmpty) {
            let (rule, pwd) = parseRule(data: item)
            rules.append(rule)
            passwords.append(pwd)
        }
    }
    var res = 0
    var i = 0
    for password in passwords {
        if (rules[i].Validate2(s: password)) {
            res += 1
        }
        i += 1
    }
    print("Valid: \(res)")
}

solve1()
solve2()

File: day2/cpp/main.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <algorithm>
#include <sstream>
#include <tuple>

using namespace std;


class PasswordRule
{
public:
	int min_occurances;
	int max_occurances;
	string pwd_char;

    bool validate(string s) const
    {
	    int occurances = std::count(s.begin(), s.end(), pwd_char[0]);
	    const auto min_ok = occurances >= min_occurances;
	    const auto max_ok = occurances <= max_occurances;
        return min_ok && max_ok;
    }
	bool validate_2(string s) const
    {
        auto ok1 = s[min_occurances - 1] == pwd_char[0];
        auto ok2 = s[max_occurances - 1] == pwd_char[0];
        auto both_ok = ok1 && ok2;
        return (ok1 || ok2) && !both_ok;
    }
};

std::vector<std::string> split(const std::string text, const char delim) {
    std::string line;
    std::vector<std::string> vec;
    std::stringstream ss(text);
    while (std::getline(ss, line, delim)) {
        vec.push_back(line);
    }
    return vec;
}

static tuple<PasswordRule, string> parse_rule(string s)
{
    auto res = PasswordRule();
    auto parts = split(s, ' ');
    auto ranges = split(parts[0], '-');
    res.min_occurances = stoi(ranges[0]);
    res.max_occurances = stoi(ranges[1]);
    res.pwd_char = parts[1].replace(parts[1].find(":"), 3, "");
    return make_tuple(res, parts[2]);
}

static void solve_part1(vector<string> raw_data)
{
    auto rules = vector<PasswordRule>();
    auto passwords = vector<string>();
	for(auto item : raw_data)
	{
        PasswordRule rule;
        string pwd;
        tie(rule, pwd) = parse_rule(item);
        rules.emplace_back(rule);
        passwords.emplace_back(pwd);
	}

    auto i = 0, res = 0;
    for(auto password : passwords)
    {
        if (rules[i].validate(password)) res++;
        i++;
    }
    cout << "Valid passwords: " << res << "\n";
}

static void solve_part2(vector<string> raw_data)
{
    auto rules = vector<PasswordRule>();
    auto passwords = vector<string>();
    for (auto item : raw_data)
    {
        PasswordRule rule;
        string pwd;
        tie(rule, pwd) = parse_rule(item);
        rules.emplace_back(rule);
        passwords.emplace_back(pwd);
    }

    auto i = 0, res = 0;
    for (auto password : passwords)
    {
        if (rules[i].validate_2(password)) res++;
        i++;
    }
    cout << "Valid passwords: " << res << "\n";
}

int main()
{
    const string filename = "c:\\temp\\aoc2.txt";
    ifstream inf(filename);
    vector<string> lines;
    string line;
    for (unsigned int i = 1; getline(inf, line); ++i)
        lines.emplace_back(line);

    solve_part1(lines);
    solve_part2(lines);
}